src/engine.ml

```diff
@@ -584,51 +594,58 @@ let incoming state now ts buf =
-             let session, transport, encs =
-               wrap_hmac_control now state'.session key ch''.transport outs
+           match
+             check_control_integrity bad_mac key p state.session.their_hmac >>= fun () ->
+             (* _first_ update state with most recent received packet_id *)
+             expected_packet state.session ch.transport d
+           with
+           | Error e ->
+             (* only in udp mode? *)
+             Logs.warn (fun m -> m "ignoring bad packet %a" pp_error e);
+             Ok (state, out, None)
+           | Ok (session, transport) ->
+             let state' = { state with session }
+             and ch' = { ch with transport }
              in
-             let out' = out @ encs in
-             let ch''' = { ch'' with transport } in
-             let state' = { state' with config = config' ; session } in
-             if est then
-               match state'.state with
-               | Handshaking _ ->
-                 let ip_config = ip_from_config config'
-                 and compress = match Config.find Comp_lzo config' with
-                   | None -> false | Some () -> true
-                 in
-                 let session = { state'.session with compress } in
-                 let mtu = mtu config' compress in
-                 let act' =
-                   match act with None -> Some (`Established (ip_config, mtu))
-                                | _ -> assert false
-                 in
-                 { state' with state = Ready ; session ; channel = ch''' }, out', act'
-               | Rekeying _ ->
-                 (* TODO: may cipher (i.e. mtu) or compress change between rekeys? *)
-                 let lame_duck = Some (state'.channel, ts) in
-                 { state' with state = Ready ; channel = ch''' ; lame_duck }, out', act
-               | _ -> assert false
-             else
+             match d with
+             | `Ack _ ->
+               (* effect already handled in expected_packet (removed acked ids from out_packets) *)
+               Logs.info (fun m -> m "ignoring acknowledgement");
+               Ok (set_ch state' ch', out, act)
+             | `Control (typ, (_, _, data)) ->
+               incoming_control state.config state.rng state'.session ch' now typ data
+               >>| fun (est, config', ch'', outs) ->
+               Logs.debug (fun m -> m "out channel %a, pkts %d"
+                              pp_channel ch'' (List.length outs));
+               (* each control needs to be acked! *)
+               let outs = match outs with [] -> [ Cstruct.empty ] | xs -> xs in
+               (* now prepare outgoing packets *)
+               let session, transport, encs =
+                 wrap_hmac_control now ts state'.session key ch''.transport outs
+               in
+               let out' = out @ encs in
+               let ch''' = { ch'' with transport } in
+               let state' = { state' with config = config' ; session } in
```

Comment:
I find it wildly confusing that we have ch-prime-prime-prime, but state-prime gets to be rebound, as does `session` (but no primes there).

Reply:
i refactored that code a bit a36ee70